src/systems/filecoin_mining/storage_mining/storage_mining_subsystem.go

```diff
@@ -281,47 +340,72 @@ func (sms *StorageMiningSubsystem_I) _verifyElection(header block.BlockHeader, o
-// 	panic("TODO") // TODO: get worker address and key pair
-
-// 	// TODO: is this just workerAddress, or is there a separation here
-// 	// (worker is AccountActor, workerMiner is StorageMinerActor)?
-// 	var workerMinerActorAddress addr.Address
-// 	panic("TODO")
-
-// 	var gasPrice msg.GasPrice
-// 	var gasLimit msg.GasAmount
-// 	panic("TODO") // TODO: determine gas price and limit
-
-// 	var callSeqNum actor.CallSeqNum
-// 	panic("TODO") // TODO: retrieve CallSeqNum from worker
-
-// 	messageParams := actor.MethodParams([]actor.MethodParam{
-// 		actor.MethodParam(poster.Serialize_PoStSubmission(postSubmission)),
-// 	})
-
-// 	unsignedMessage := msg.UnsignedMessage_Make(
-// 		workerAddress,
-// 		workerMinerActorAddress,
-// 		Method_StorageMinerActor_SubmitPoSt,
-// 		messageParams,
-// 		callSeqNum,
-// 		actor.TokenAmount(0),
-// 		gasPrice,
-// 		gasLimit,
-// 	)
-
-// 	signedMessage, err := msg.Sign(unsignedMessage, workerKeyPair)
-// 	if err != nil {
-// 		return err
-// 	}
-
-// 	err = sms.FilecoinNode().SubmitMessage(signedMessage)
-// 	if err != nil {
-// 		return err
-// 	}
-
-// 	return nil
-// }
+func (sms *StorageMiningSubsystem_I) _trySurprisePoSt(currState stateTree.StateTree, sma StorageMinerActorState) sector.OnChainPoStVerifyInfo {
+
+	// get randomness for SurprisePoSt
+	challEpoch := sma.ChallengeStatus().LastChallengeEpoch()
+	randomnessK := sms._consensus().GetPoStChallengeRand(sms._blockchain().BestChain(), challEpoch)
+	input := sms.PreparePoStChallengeSeed(randomnessK, sms._keyStore().MinerAddress())
+	postRandomness := sms._keyStore().WorkerKey().Impl().Generate(input).Output()
+
+	// TODO: add how sectors are actually stored in the SMS proving set
+	util.TODO()
+	provingSet := make([]sector.SectorID, 0)
+
+	candidates := sms.StorageProving().Impl().GenerateSurprisePoStCandidates(postRandomness, provingSet)
+
+	if len(candidates) <= 0 {
+		// Error. Will fail this surprise post and must then redeclare faults
+		return nil // fail to generate post candidates
+	}
+
+	winningCandidates := make([]sector.PoStCandidate, 0)
+	for _, candidate := range candidates {
+		if sma._verifySurprisePoStMeetsTargetReq(candidate) {
+			winningCandidates = append(winningCandidates, candidate)
+		}
+	}
+
+	postProof := sms.StorageProving().Impl().CreateSurprisePoStProof(postRandomness, winningCandidates)
+
+	var ctc sector.ChallengeTicketsCommitment // TODO: proofs to fix when complete
+	surprisePoSt := &sector.OnChainPoStVerifyInfo_I{
+		CommT_:      ctc,
+		Candidates_: winningCandidates,
+		Randomness_: postRandomness,
+		Proof_:      postProof,
+	}
+	return surprisePoSt
+}
+
+func (sms *StorageMiningSubsystem_I) _submitSurprisePoStMessage(state stateTree.StateTree, sPoSt sector.OnChainPoStVerifyInfo, gasPrice actor.TokenAmount, gasLimit msg.GasAmount) error {
+
+	params := make([]util.Serialization, 1)
+	params[0] = sector.Serialize_OnChainPoStVerifyInfo(sPoSt)
+
+	sysActor, ok := state.GetActor(addr.SystemActorAddr)
+	Assert(ok)
```

Comment:
System actor is not involved here. The sender is the worker account and must use that sequence number.